test/test_udacity.py

```diff
@@ -0,0 +1,836 @@
+    y = th.tensor([1, 1, 1, 1, 1])
+
+    # throws error
+    # z = x + y
+
+    x = th.tensor([1, 2, 3, 4, 5]).send(bob)
+    y = th.tensor([1, 1, 1, 1, 1]).send(alice)
+
+    # throws error
+    # z = x + y
+
+    from torch import nn, optim
+
+    # A Toy Dataset
+    data = th.tensor([[1., 1], [0, 1], [1, 0], [0, 0]], requires_grad=True)
+    target = th.tensor([[1.], [1], [0], [0]], requires_grad=True)
+
+    # A Toy Model
+    model = nn.Linear(2, 1)
+
+    opt = optim.SGD(params=model.parameters(), lr=0.1)
+
+    def train(iterations=20):
+        for iter in range(iterations):
+            opt.zero_grad()
+
+            pred = model(data)
+
+            loss = ((pred - target) ** 2).sum()
+
+            loss.backward()
+
+            opt.step()
+
+            print(loss.data)
+
+    train()
+
+    data_bob = data[0:2].send(bob)
+    target_bob = target[0:2].send(bob)
+
+    data_alice = data[2:4].send(alice)
+    target_alice = target[2:4].send(alice)
+
+    datasets = [(data_bob, target_bob), (data_alice, target_alice)]
+
+    def train(iterations=20):
+
+        model = nn.Linear(2, 1)
+        opt = optim.SGD(params=model.parameters(), lr=0.1)
+
+        for iter in range(iterations):
+
+            for _data, _target in datasets:
+                # send model to the data
+                model = model.send(_data.location)
+
+                # do normal training
+                opt.zero_grad()
+                pred = model(_data)
+                loss = ((pred - _target) ** 2).sum()
+                loss.backward()
+                opt.step()
+
+                # get smarter model back
+                model = model.get()
+
+                print(loss.get())
+
+    train()
+
+    bob.clear_objects()
+    alice.clear_objects()
+
+    x = th.tensor([1, 2, 3, 4, 5]).send(bob)
+
+    x = x.send(alice)
+
+    bob._objects
+
+    alice._objects
+
+    y = x + x
+
+    y
+
+    bob._objects
+
+    alice._objects
+
+    jon = sy.VirtualWorker(hook, id="jon")
+
+    bob.clear_objects()
+    alice.clear_objects()
+
+    x = th.tensor([1, 2, 3, 4, 5]).send(bob).send(alice)
+
+    bob._objects
+
+    alice._objects
+
+    x = x.get()
+    x
+
+    bob._objects
+
+    alice._objects
+
+    x = x.get()
+    x
+
+    bob._objects
+
+    bob.clear_objects()
+    alice.clear_objects()
+
+    x = th.tensor([1, 2, 3, 4, 5]).send(bob).send(alice)
+
+    bob._objects
+
+    alice._objects
+
+    del x
+
+    bob._objects
+
+    alice._objects
+
+    bob.clear_objects()
+    alice.clear_objects()
+
+    x = th.tensor([1, 2, 3, 4, 5]).send(bob)
+
+    bob._objects
+
+    alice._objects
+
+    x.move(alice)
+
+    bob._objects
+
+    alice._objects
+
+    x = th.tensor([1, 2, 3, 4, 5]).send(bob).send(alice)
+
+    bob._objects
+
+    alice._objects
+
+    x.remote_get()
+
+    bob._objects
+
+    alice._objects
+
+    x.move(bob)
+
+    x
+
+    bob._objects
+
+    alice._objects
+
+    assert True
+
+
+def test_section_3_securing_fl(hook):
```

Comment:
Should you have assertions in this code

Reply:
no - it's just checking that the code runs